### ISSUE TITLE
Sharing is caring

### ISSUE THREAD

**Issue by javi11** [javi11/usenet-drive#160]
# Future ideas to reduce the space taken on usenet network and improve the upload speed.

1. Every user using usenet-drive will share a p2p network
2. On uploading something, a unique hash will be generated before encrypting
3. This hash will be used to check if someone on the network has the same file aka (nzb)
4. If someone has it it will simply downloaded from that user and the file won't be uploaded to usenet

# Considerations

- The files will be encrypted with a random password, since the nzb will be shared with the password we don't want the user to choose a password of their own.
- The only way to get an existing nzb from the network will be if the user already have the file and performs an upload to the usenet network, this way everything will be private and you won't be able to potencially scan other users

# Benefits
- Uploads can be faster
- We don't saturate the usenet network losing our unlimited benefits

**Comment by javi11** [comment]
Good questions @MythodeaLoL, so far I don't have the full idea on how to do it jajaja but here are my comments to your questions:

`Could this hash/md5 already be implemented currently, to support future updates?`
Yes I want to include the hash md5 + encrypt sooner to prepare the ground, right now it won't be possible since we are using rclone crypt and since we are using different passwords the content will be different.

`files already uploaded, what would they look like in the face of these changes?`
My idea is create a migration script to add the rclone crypt password to those nzbs and to create the hash for the actual content

`Would users who did not wish to enable p2p networking and hash sharing have this choice?`

Yes ofc all will be optional, but my idea is enable it by default

`Where would the central server be hosted to connect p2p clients?`

No idea yet ajaj

`Would it be possible to configure self-hosted central servers, for access to specific people?`

Probably

`And if, for some reason, the articles already sent and in the p2p database are corrupted, how would this be handled? since it would always get a hash/nzb from p2p that is corrupt.`

I will need to improve the corrupted detection to ignore those files

**Comment by javi11** [comment]
M this will be assuming you share the google drive with others using it no? I think this is an interesting feature that can be achieved now but with a different focus, like save the hash on the nzb and if you try to upload something repeatly it will simple give the same nzb. This way you can mount your google drive using rclone for instance an the use it  as your nzbs folder

**Comment by MythodeaLoL** [comment]
Not really, I thought that over time this could generate a large amount of data that could take up a lot of space on the HD. The use of a cloud server would be more beneficial to the local system.

What you're talking about is the old way I used it, currently I store them all on a central server, without synchronization with the cloud, but synchronized locally with my PC, all my friends who need it only access it via the webdav url. This way everything is concentrated in one place, making maintenance easier.

I believe that in this case, if a user wanted to maintain the p2p server, and also the Usenet-Drive server, he would have to somehow have a copy of everything he has in the /nzbs folder, for example on the p2p server.

That's why I thought that data redundancy, however serving different means p2p network(hash nzbs) and usenet-drive network(local nzbs) could result in a large use of disk space, so I suggested using the cloud as a better alternative, but this would depend exclusively on the Host user.

I think I wasn't very clear in my explanation before. haha :)
